torch/optim/lbfgs.py

```diff
@@ -239,3 +252,136 @@ def step(self, closure):
         state['prev_loss'] = prev_loss
 
         return orig_loss
+
+    def _copy_param(self):
+        original_param_data_list = []
+        for p in self._params:
+            param_data = p.data.new(p.size())
+            param_data.copy_(p.data)
+            original_param_data_list.append(param_data)
+        return original_param_data_list
+
+    def _set_param(self, param_data_list):
+        for i in range(len(param_data_list)):
+            self._params[i].data.copy_(param_data_list[i])
+
+    def _set_param_incremental(self, alpha, d):
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            p.data.copy_(p.data + alpha*d[offset:offset + numel].resize_(p.size()))
+            offset += numel
+        assert offset == self._numel()
+
+    def _directional_derivative(self, d):
+        deriv = 0.0
+        offset = 0
+        for p in self._params:
+            numel = p.numel()
+            deriv += torch.sum(p.grad.data * d[offset:offset + numel].resize_(p.size()))
+            offset += numel
+        assert offset == self._numel()
+        return deriv
+
+    def _max_alpha(self, d):
```

Comment:
Reference for Cauchy point computation: equation 16.71 in Numerical Optimization 2nd Edition by Nocedal and Wright.